src/main/java/net/ripe/rpki/validator3/domain/RpkiObjectCleanupService.java

```diff
@@ -0,0 +1,155 @@
+
+import javax.persistence.EntityManager;
+import javax.persistence.FlushModeType;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+@Service
+@Slf4j
+public class RpkiObjectCleanupService {
+
+    @Autowired
+    private EntityManager entityManager;
+
+    @Autowired
+    private TrustAnchors trustAnchors;
+
+    @Autowired
+    private RpkiObjects rpkiObjects;
+
+    @Autowired
+    private TransactionTemplate transactionTemplate;
+
+    private final Duration cleanupGraceDuration;
+
+    public RpkiObjectCleanupService(@Value("${rpki.validator.rpki.object.cleanup.grace.duration}") String cleanupGraceDuration) {
+        this.cleanupGraceDuration = Duration.parse(cleanupGraceDuration);
+    }
+
+    /**
+     * Marks all RPKI objects that are reachable from a trust anchor by following the entries in the manifests.
+     * Objects that are no longer reachable will be deleted after a configurable grace duration.
+     */
+    @Scheduled(initialDelay = 60_000, fixedDelayString = "${rpki.validator.rpki.object.cleanup.interval.ms}")
+    public long cleanupRpkiObjects() {
+        Instant now = Instant.now();
+        for (TrustAnchor trustAnchor : trustAnchors.findAll()) {
+            transactionTemplate.execute((status) -> {
+                entityManager.setFlushMode(FlushModeType.COMMIT);
+
+                log.info("tracing objects for trust anchor {}", trustAnchor);
+
+                X509ResourceCertificate resourceCertificate = trustAnchor.getCertificate();
+                if (resourceCertificate != null) {
+                    traceCertificateAuthority(now, resourceCertificate);
+                }
+
+                return null;
+            });
+        }
+
+        return deleteUnreachableObjects(now);
+    }
+
+    private long deleteUnreachableObjects(Instant now) {
+        return transactionTemplate.execute((status) -> {
+            entityManager.flush();
+
+            Instant unreachableSince = now.minus(cleanupGraceDuration);
+            long count = rpkiObjects.deleteUnreachableObjects(unreachableSince);
+            log.info("Deleted {} RPKI objects that have not been marked reachable since {}", count, unreachableSince);
+            return count;
+        });
+    }
+
+    private void traceCertificateAuthority(Instant now, X509ResourceCertificate resourceCertificate) {
+        if (resourceCertificate == null || resourceCertificate.getManifestUri() == null) {
+            return;
+        }
+
+        Optional<RpkiObject> maybeManifest = rpkiObjects.findLatestByTypeAndAuthorityKeyIdentifier(
+            RpkiObject.Type.MFT,
+            resourceCertificate.getSubjectKeyIdentifier()
+        );
+        maybeManifest.ifPresent(manifest -> {
+            markAndTraceObject(now, "manifest.mft", manifest);
+        });
+    }
+
+    private void markAndTraceObject(Instant now, String name, RpkiObject rpkiObject) {
+        if (now == rpkiObject.getLastMarkedReachableAt()) {
```

Comment:
Is it a good idea to compare timestamps? 
Could there be inequality caused by inaccuracies?

Reply:
I actually use object instance identity here to see if an object has been marked during the current run. I'll add a comment, but it might be clearer (but slower) to have a special `visitedObjects` set.